internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -100,159 +111,217 @@ func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineD
-	newRevision := strconv.FormatInt(maxOldRevision+1, 10)
+	newRevisionInt := maxOldRevision + 1
+	newRevision := strconv.FormatInt(newRevisionInt, 10)
 
-	// Latest machine set exists. We need to sync its annotations (includes copying all but
-	// annotationsToSkip from the parent deployment, and update revision, desiredReplicas,
-	// and maxReplicas) and also update the revision annotation in the deployment with the
-	// latest revision.
+	// Latest MachineSet exists. We need to sync all the in-place propagated values.
 	if existingNewMS != nil {
-		msCopy := existingNewMS.DeepCopy()
-		patchHelper, err := patch.NewHelper(msCopy, r.Client)
-		if err != nil {
-			return nil, err
-		}
-
-		// Set existing new machine set's annotation
-		annotationsUpdated := mdutil.SetNewMachineSetAnnotations(d, msCopy, newRevision, true, log)
-
-		minReadySecondsNeedsUpdate := msCopy.Spec.MinReadySeconds != *d.Spec.MinReadySeconds
-		deletePolicyNeedsUpdate := d.Spec.Strategy.RollingUpdate.DeletePolicy != nil && msCopy.Spec.DeletePolicy != *d.Spec.Strategy.RollingUpdate.DeletePolicy
-		if annotationsUpdated || minReadySecondsNeedsUpdate || deletePolicyNeedsUpdate {
-			msCopy.Spec.MinReadySeconds = *d.Spec.MinReadySeconds
-
-			if deletePolicyNeedsUpdate {
-				msCopy.Spec.DeletePolicy = *d.Spec.Strategy.RollingUpdate.DeletePolicy
+		// FIXME: In the current implementation any changes to in-place propagated fields do not cause the revision to change.
+		// Therefore we loose track of the history of MachineDeployment changes and therefore it wont be able to revert back
+		// to an old state (a state that differs only in the in-place propagated field values).
+		// In the next iteration this will be changed so that in-place propagated values are also captures in history while
+		// not triggering a rollout.
+
+		// The existingMS's revision should be the greatest among all MSes. Usually, its revision number is newRevision
+		// (the max revision number of all old MSes + 1). However, it's possible that some old MSes are deleted
+		// after the existingMS's revision was updated, and newRevision becomes smaller than the current existingMS's
+		// revision. We should only update existingMS's revision when it's smaller than newRevision.
+		// TODO: check if the above comment is actually correct. (The comment was copied from old code). Can the case actually happen?
+		oldRevision, ok := existingNewMS.Annotations[clusterv1.RevisionAnnotation]
+		if ok {
+			oldRevisionInt, err := strconv.ParseInt(oldRevision, 10, 64)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse oldRevision on MachineSet")
+			}
+			if newRevisionInt < oldRevisionInt {
+				newRevision = oldRevision
 			}
+		}
 
-			return nil, patchHelper.Patch(ctx, msCopy)
+		desiredMS, err := r.computeDesiredMachineSet(d, existingNewMS, oldMSs, newRevision)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to compute desired MachineSet")
+		}
+		patchOptions := []client.PatchOption{
```

Comment:
@fabriziopandini Do you remember why we required allowedPaths in CC reconcilation?

I think one reason was to not take ownership of the entire Cluster object. But I think here we should be fine without adding something like this

Reply:
I looked into what we are sending here with SSA to the server and the only thing which I didn't expect was `creationTimestamp: null`. Not sure if that's a problem though given it can't really be set to null.

Reply:
> @fabriziopandini Do you remember why we required allowedPaths in CC reconcilation?

Because some fields are set by other controllers (e.g control plane endpoint) and thus we must not include them in the intent/in the set of fields owned by the topology controller.

Reply:
Okay then we should be good in this case as we are calculating the MachineSet from scratch and not building on top of one that we get from the server (like in computeCluster in the CC reconciler)